racket/collects/racket/private/for.rkt

```diff
@@ -1801,43 +1828,54 @@
                            [delayed-id delayed-id]
                            [delayer-id delayer-id])
                (check-identifier-bindings #'orig-stx #`(fold-var ... delayed-id) "accumulator" (void))
-                (cond
-                  [right?
-                   (define loop-stx
-                     (quasisyntax/loc #'orig-stx
-                       (for/foldX/derived [orig-stx inner-recur/foldr #,for*? #f ()]
-                         ()
-                         (done-k-proc)
-                         (done-k-proc)
-                         #f
-                         . rest)))
-                   (quasisyntax/loc #'orig-stx
-                     (let ()
-                       (define (done-k-proc) (#%expression (values* fold-init ...)))
-                       (define-syntax inner-recur/foldr
-                         #,(if delay?
-                               #'(make-inner-recur/foldr/lazy
-                                  (list (quote-syntax fold-var) ...)
-                                  (quote-syntax delayed-id)
-                                  (quote-syntax delayer-id))
-                               #'(make-inner-recur/foldr/strict
-                                  (list (quote-syntax fold-var) ...))))
-                       #,(if result-expr
-                             ;; Make sure `fold-var`s in `result-expr` are also delayed, if relevant
-                             #`(inner-recur/foldr () [#,result-expr] #,loop-stx)
-                             loop-stx)))]
-                  [else
-                   (define loop-stx
-                     (quasisyntax/loc #'orig-stx
-                       (for/fold/derived/final [orig-stx #,for*?]
-                         ([fold-var fold-init] ...)
-                         (values* fold-var ...)
-                         . rest)))
-                   (if result-expr
-                       (quasisyntax/loc #'orig-stx
-                         (let-values ([(fold-var ...) #,loop-stx])
-                           #,result-expr))
-                       loop-stx)])))]
+               (with-syntax ([(int-var ...)
+                              (map (lambda (fold-var)
+                                     (datum->syntax fold-var (gensym (syntax-e fold-var))))
```

Comment:
I think the problem is just that we need to copy the syntax properties from `fold-var` onto `int-var` (and that fixes my initial testing). Is that generally reasonable?

Reply:
Can you send your change that made things work? I tried
```
(datum->syntax fold-var (gensym (syntax-e fold-var)) fold-var fold-var)
```
in place of
```
(datum->syntax fold-var (gensym (syntax-e fold-var)))
```
but was not able to build the `math` collection after that change.